File: 503. Next Greater Element II.go
func nextGreaterElements(nums []int) []int {
    n := len(nums)
    answer := make([]int, n)
    for i := range answer {
        answer[i] = -1
    }
    stack := make([]int, 0, n)
    for i := 0; i < 2*n; i++ {
        i := i % n
        for len(stack) > 0 && nums[stack[len(stack)-1]] < nums[i] {
            index := stack[len(stack)-1]
            stack = stack[:len(stack)-1]
            if index >= len(answer) { continue }
            answer[index] = nums[i]
        }
        stack = append(stack, i)
    }
    return answer
}

File: 503. Next Greater Element II.java
class Solution {
    public int[] nextGreaterElements(int[] nums) {
        int n = nums.length;
        int[] answer = new int[nums.length];
        Arrays.fill(answer, -1);
        Deque<Integer> stack = new ArrayDeque<>();
        for (int i = 0; i < 2*nums.length; i++) {
            while (!stack.isEmpty() && nums[stack.peekLast()%n] < nums[i%n]) {
                int index = stack.pollLast();
                if (index >= answer.length) continue;
                answer[index] = nums[i%n];
            }
            stack.addLast(i);
        }
        return answer;
    }
}
